### ISSUE TITLE
Task Notifications

### ISSUE THREAD

**Issue by Rybec** [Rybec/pyRTOS#4]
FreeRTOS has task notifications: https://www.freertos.org/RTOS-task-notifications.html

These are essentially triggers a task can block on, that another task can trigger to unblock the task.  Task notifications in FreeRTOS have an associated 32 bit value that the triggering task can set or increment.  A task can only block on one notification at a time, but tasks can have multiple notifications.  (The pyRTOS blocking mechanism allows for blocking on multiple conditions, _but_ you won't know which one you unblocked on without checking them all.)  In FreeRTOS the number of tasks is set by a constant, but in Python we could set this globally or task-by-task, at runtime.

What are the advantages of notifications over messages?  I've been wanting to make light tasks that don't have a built in incoming message queue.  Notifications could provide a lighter message passing mechanic for light tasks.  It might even be worth making light, notification-only tasks the default, and have tasks with built in mailboxes as a heavier alternative.  The big difference is that messages have a return address, while notifications do not.

We might even be able to use an inheritance mechanism to allow users to create tasks with _only_ the features they need.

Anyhow, the question here is, is there really a need for task notifications?  They shouldn't be hard to implement, but if there isn't much interest, I will prioritize other features.

(Also, since this is intended for microcontrollers with limited resources, I don't want to add a lot of feature that will take up space but rarely see use.  FreeRTOS can get away with it, because it is written in a compiled language, where the compiler/linker can skip parts that are not used in the immediate application.  We can't do that with CircuitPython.)

**Comment by Rybec** [comment]
Still undecided, but task notifications could be really useful for Service Routines.  Constructing Messages or using MessageQueues is significantly more expensive in memory and processing than well designed task notifications would be.

The question is: What applications could task notifications have, in the context of service routines?

One answer is: GPIO pin interrupts could be simulated, where a Service Routine is used to poll the pins and then, when there is a state change, it sends task notifications to any tasks that need to know about it.  This would be lighter than sending messages and cleaner than using a global variable, and it would allow more precise blocking for tasks waiting on such a state change.

Are there other use cases for task notifications?  Perhaps notifying tasks of waiting data in buffers?  Some of this could be managed with custom blocking conditions, but maybe this would provide a cleaner, more standard option, that could be used with OS extensions for things like networking, pin interrupts, and perhaps even timer alarms (beyond just delays).

**Comment by Rybec** [comment]
Let's do this.  The completion of issue #5 has created a greater need for these.  The very light weight communication of task notifications is now necessary to provide for communication that can keep service routines as light as possible.

**Comment by Rybec** [comment]
Ok, so new question:

1. Are we going to implement a new kind of task that has task notifications?
2. Are we going to add task notifications to the current `Task` class?
3. Are we going to split out the mailbox capability from our `Task` class and then create class extensions that can be used to add mailboxes and notifications to tasks?

I'm leaning toward number 3.  I wonder if there is a simple way of doing this using decorators.  I don't want users to have to implement custom `Task` classes that inherit from `Task` and mailbox and notification objects (essentially implementing mailboxes and notifications as equivalent to Java interfaces), if I can avoid it.  If this is possible to do easily with decorators, let's do that with number 3.

Side note: The next release is going to be pyRTOS 1.0.0.  Service routines haven't broken userspace, but if I do number 3 above, this definitely will.

**Comment by Rybec** [comment]
Task Notification Design:

A `Task` with notifications will have a tuple of two array.arrays.  The first array will contain signed bytes, and the second will contain signed longs.  The length of the arrays will depend on the notification types desired.  A notification tuple with 4 elements would be initialized like this: `(array.array('b', [0, 0, 0, 0], array.array('l', [0, 0, 0, 0]))`.

The first array, with bytes, is notification flags.  0 means there is no notification pending, and the task has no interest in the current value of the notification.  Use higher values as needed.  For example, 2 might mean the notification has been sent but the task has not read it, and 1 might mean the task has read the value but is still interested in it.  Alternatively, the flag value might be used to signal urgency or priority.  (Functions for setting and reading notifications should have an option for automatically incrementing, decrementing, or setting the flag level.)

The second array is the message associated with the notification.  Not all notifications will need messages.  Sometimes just the flag is enough, and since the flag has 8 bits to work with, there's a lot of potential room for communication with just that.  But, the second array provides 32 signed bits for data, so you can keep data separate from the flag and have more room for data in the communication.  This can be used to communicate specific data, or it can be used as a counter or an index.


I should be able to use decorator functions to dynamically add notification and/or mailbox capabilities to tasks.  I am not sure this is cheaper than using inheritance, but it will be easier to use, and it may be cheaper when we take into account potential permutations of task customization.  Right now, we only have two "decorations" to implement, notifications and mailboxes.  This gives us four permutations: One, the other, both, or neither.  What happens if we add another one?  Now we have 8 options.  And a fourth gives us 16 options.  Inheritance based construction becomes exponentially more expensive, in class metadata, unless applications only ever use the same task build for all tasks.  Python's dynamic typing means that we can add both attributes and methods to task dynamically, avoiding the need to use flash space to define custom task objects.  (I wonder how much space class definitions take in compiled .mpy files...)

I should probably check, but it's possible dynamically added functions are more expensive in memory than functions defined as part of a class.  If this is true, then decorators are probably a bad strategy for this...  Is there some kind of memory profiling I can do to check this?  I guess I am still going to have to think on this before going ahead with implementation.

**Comment by Rybec** [comment]
I decided to implement this as an optional element selected on object instantiation, using a keyword argument.  I also changed the built in message queues (now called "mailboxes" in the documentation) to default to off but be selectable using a keyword argument.

Anyhow, this feature has been added, however it is currently untested.  I intend on updating the Rotary Trinkey test to use notifications to send color data to the rendering task.  (The new mailbox code is tested in sample.py.)  Closing this issue now!
